### PR TITLE
feat: nested selection conflict detection

### PR DESCRIPTION
**Pull request by evpeople**
I implemented conflict path detection using HashMap. Using a prefix tree might also be a worthwhile option to consider.

Closes https://github.com/sxyazi/yazi/issues/688

### CODE REVIEW

**Comment by sxyazi** [comment]
Thank you for the PR!

Could you also add an `insert_many(&mut self, urls: &[&Url], parent: Url)` method? This is very useful for bulk selecting (selecting all) files - all of which have the same parent path, so it only needs to be detected once to improve performance.

Then `insert()` can simply be a wrapper for `insert_many()` like this:

```rust
	pub fn insert(&mut self, url: &Url) -> bool {
            self.insert_many(&[url], url.parent_url())
        }
```

**Comment by evpeople** [comment]
Ok, I will try it

**Comment by evpeople** [comment]
Considering that the paths corresponding to insert_many all have the same parent directory, 
I implemented insert_many by modifying the logic in the original insert method that checks the object being inserted to instead check the first object in insert_many.

I didn't use the parent variable. I tried some methods that involved the parent variable, but they were not successful in passing the tests.

**Comment by evpeople** [comment]
To make the compilation pass, I added contains and get_inner methods. I am unsure if using the BTree inside inner within yazi-fm might lead to inconsistencies between HashMap and BTree.

File: yazi-core/src/tab/selected.rs [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+	}
+	pub fn new() -> Self { Selected { inner: BTreeSet::new(), parents: HashMap::new() } }
+
+	pub fn insert(&mut self, url: Url) -> bool { self.insert_many(&[&url]) }
+
+	pub fn insert_many(&mut self, urls: &[&Url]) -> bool {
+		if urls.is_empty() {
+			return false;
+		}
+		let url_buf = urls[0].to_path_buf();
```
Comment (evpeople) [inline yazi-core/src/tab/selected.rs:23]: I simply changed the `url` in this line to `urls[0]`

File: yazi-core/src/tab/selected.rs [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+
+		let mut current_path = url_buf.clone();
+		let len_of_urls = urls.len();
+		loop {
+			current_path = match current_path.parent() {
+				Some(parent) => parent.to_path_buf(),
+				None => break,
+			};
+			let counter = self.parents.entry(current_path.clone()).or_insert(0);
+			*counter += len_of_urls;
```
Comment (evpeople) [inline yazi-core/src/tab/selected.rs:45]: Additionally, in this line, I changed from adding 1 to adding the length of `urls`.

File: yazi-core/src/tab/selected.rs (hunk 1 above)
Comment (sxyazi) [reply yazi-core/src/tab/selected.rs:23]: Good idea!

File: yazi-core/src/tab/selected.rs [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+		if !self.inner.remove(url) {
+			return false;
+		}
+
+		let mut current_path = url.clone();
+		while let Some(parent) = current_path.parent_url() {
+			current_path = parent;
+
+			let counter = self.parents.entry(current_path.clone()).or_insert(0);
+			*counter -= 1;
```
Comment (sxyazi) [inline yazi-core/src/tab/selected.rs:53]: When `current_path` does not exist, inserting `0` and then `0 -= 1;` will cause an overflow.

**Comment by evpeople** [comment]
The usage of Deref here has taught me a lot.

**Comment by evpeople** [comment]
I have considered the issue of data overflow. Later, I realized that if the add operation is correct, then the count in the parent can safely be decremented to 0. If overflow occurs, it indicates the possibility of some concurrency-related modifications. At this point, the program should panic.

File: yazi-core/src/tab/selected.rs [hunk 4]
```diff
@@ -36,7 +40,7 @@ impl Selected {
 			*self.parents.entry(u).or_insert(0) += urls.len();
 		}
 
-		self.inner.extend(urls.iter().cloned().cloned());
+		self.inner.extend(urls.iter().map(|&u| u.clone()));
```
Comment (evpeople) [inline yazi-core/src/tab/selected.rs:43]: I've tried many ways to get this to compile here, thank you very much for your approach, it's been very enlightening for me.

**Comment by sxyazi** [comment]
Hi, I made some little changes, and I believe it's all good now.

Thank you very much for contributing to this feature, let me merge it!

**Pull request merged**